modules/core/test/src/smithy4s/schema/SchemaPartitionSpec.scala

```diff
@@ -0,0 +1,222 @@
+    // can't possibly be decoded from.
+    val documentX = Document.obj("x" -> Document.fromInt(1))
+    val documentY = Document.obj("y" -> Document.fromInt(1))
+
+    import SchemaPartition._
+    (xPartialSchema) match {
+      case (SplittingMatch(xSchema, ySchema)) =>
+        val decoderX = Document.Decoder.fromSchema(xSchema)
+        val decoderY = Document.Decoder.fromSchema(ySchema)
+
+        val result =
+          (decoderX.decode(documentX), decoderY.decode(documentY)).mapN {
+            PartialData.unsafeReconcile(_, _)
+          }
+
+        assertEquals(result, Right(Foo(1, Some(1))))
+
+      case _ => fail("Expected partial matches")
+    }
+  }
+
+  test(
+    "Payload-specialised partial schemas can be extracted from structure schemas"
+  ) {
+    case class Foo(x: List[Int], y: Int)
+    val schema = struct(
+      list(int).required[Foo]("x", _.x),
+      int.required[Foo]("y", _.y)
+    )(Foo.apply)
+
+    // We're splitting the schema into various components. The first one is extracted
+    // as a "payload" partial, which means that schema held by the first field
+    // matching the predicate will be used as if it was a top level schema
+    val xPartialSchema = schema.payloadPartition(_.label == "x")
+
+    val documentX = Document.array(Document.fromInt(1), Document.fromInt(2))
+    val documentY = Document.obj("y" -> Document.fromInt(1))
+
+    import SchemaPartition._
+    xPartialSchema match {
+      case SplittingMatch(xSchema, ySchema) =>
+        val decoderX = Document.Decoder.fromSchema(xSchema)
+        val decoderY = Document.Decoder.fromSchema(ySchema)
+
+        val result =
+          (decoderX.decode(documentX), decoderY.decode(documentY)).mapN {
+            PartialData.unsafeReconcile(_, _)
+          }
+
+        assertEquals(result, Right(Foo(List(1, 2), 1)))
+
+      case _ => fail("Expected partial matches")
+    }
+  }
+
+  test(
+    "Computing a partial schema using a predicate that matches all fields returns the original schema"
+  ) {
+    case class Foo(x: Int, y: Option[Int])
+    val schema = struct(
+      int.required[Foo]("x", _.x),
+      int.optional[Foo]("y", _.y)
+    )(Foo.apply)
+
+    val partialSchema = schema.partition(_ => true)
+    assert(partialSchema == SchemaPartition.TotalMatch(schema))
+  }
+
+  test(
+    "Computing a partial schema using a predicate that matches no field returns no schema"
+  ) {
+    case class Foo(x: Int, y: Option[Int])
+    val schema = struct(
+      int.required[Foo]("x", _.x),
+      int.optional[Foo]("y", _.y)
+    )(Foo.apply)
+
+    val partialSchema = schema.partition(_.label.startsWith("z"))
+    assertEquals(partialSchema, SchemaPartition.NoMatch[Foo]())
+  }
+
+  test(
+    "Computing a partial schema from a non-struct schema returns no-schema"
+  ) {
+    val schema = list(int)
+
+    val partialSchema = schema.partition(_.label.startsWith("z"))
+    assertEquals(partialSchema, SchemaPartition.NoMatch[List[Int]]())
+  }
```

Comment:
**suggestion:** imo these three tests should go first as they're by far the easiest to understand, and might be a smoother introduction

Reply:
Addressed in https://github.com/disneystreaming/smithy4s/pull/875/commits/9cb8c6bde034946d6d0332034e385716cf15b9ee